### ISSUE TITLE
Central tracker: model benchmark failures across GDP strategies

### ISSUE THREAD

**Issue by tristantc** [SECQUOIA/gdplib#60]
This is the central tracker for model-level benchmark failures across GDP transformations and GDPopt solution strategies.

PR #58 should stay scoped to benchmark infrastructure: runner behavior, solver metadata, JSON output, generated artifacts, and summary tooling. Model formulation fixes should be tracked here and implemented through focused model PRs.

Status synchronized with the cleaned PR #58 description on 2026-05-08.

## Scope

Track and coordinate fixes for models that do not currently work with every intended strategy when using the current benchmark path with GAMS/BARON subsolvers.

Out of scope for this issue:
- PyPI/release workflow work: #104 and #51.
- General solution-record schema work: #105, except where a model fix establishes verified objective metadata.
- Documentation-only source audits, except when they are needed to justify a model behavior change. Methanol units/constants are tracked separately in #103.

Related issues:
- #55: Pyomo.GDP and GDPopt deprecations. This should be handled before or alongside the first model-fix PRs because it affects boolean/indicator initialization and GDPopt option names.
- #42: methanol infeasibility investigation. Coordinate with #70.
- #44: Kaibel `gdp.hull` `ZeroDivisionError`. Coordinate with #68.
- #46: multiple formulations/instances handling, closed but relevant for verification of `mod_hens`, `modprodnet`, and `stranded_gas`.
- #30, #31, #32, #33, #34, #35: closed historical benchmark/model issues. Useful lessons: hull needs finite variable bounds inside disjuncts, `build_model()` must be solver-free and return a model, and solver/environment failures must be separated from model formulation failures.

## Working Rules

Each model-fix PR should:
- reference this issue and the relevant per-model issue;
- keep `build_model()` import/model construction solver-free;
- preserve source references and benchmark intent;
- add or update focused tests for import, construction, and any transformed model path that the fix changes;
- include `gdp.bigm` and/or `gdp.hull` transformation smoke tests when those strategies are changed;
- keep solver-backed verification separate from default tests unless the solver is already available in CI;
- report exact benchmark commands, Pyomo/GAMS/BARON versions, solver interface, termination condition, objective value, gap, and any residual infeasibility evidence in the PR or issue comment;
- avoid committing generated `benchmark_result` or `benchmark_summary` outputs.

## Recommended Sequence

1. Stabilize shared tooling before model PRs.
   - Merge the benchmark infrastructure work from PR #58 or rebase model PRs on it.
   - Address #55 in a small cross-cutting PR: boolean/indicator values should use `True`/`False`, and GDPopt should use the current `algorithm` option where applicable.
   - Run a fresh benchmark/reproduction pass with the updated runner and update each per-model issue with current failure logs. This prevents fixing stale failures from the old benchmark script.

2. Tackle the highest-blast-radius model failures first.
   - #68 `kaibel`, with #44, because every listed strategy is still open and there is a concrete `gdp.hull` division-by-zero failure.
   - #76 `stranded_gas`, because six strategies remain and it has multiple-instance/formulation concerns from #46.
   - #65 `ex1_linan_2023`, because five GDPopt/transformation strategies remain.
   - #62 `biofuel`, because hull/enumeration/GLOA/LBB remain and it previously exposed Big-M suffix/API issues in #31.
   - #66 `gdp_col`, because hull/GLOA/LBB failures may share bounded-variable/domain lessons with #68.

3. Work through medium-width model failures.
   - #67 `hda`: hull and GLOA.
   - #71 `mod_hens`: enumeration, GLOA, and LBB. Verify all supported formulations/instances from #46.
   - #69 `med_term_purchasing`: enumeration and LBB.
   - #75 `spectralog`: enumeration and GLOA.
   - #77 `syngas`: GLOA and LBB.

4. Sweep the likely common GDPopt/GLOA-only failures after the shared GDPopt configuration is fixed.
   - #61 `batch_processing`: GLOA.
   - #63 `cstr`: GLOA.
   - #72 `modprodnet`: GLOA. Verify all supported formulations/instances from #46.
   - #73 `positioning`: GLOA.
   - #74 `small_batch`: GLOA.
   - #70 `methanol`: GLOA, but coordinate separately with #42 and #103 because there is a documented infeasibility/initialization/constants investigation.

5. Finish with benchmark documentation and result metadata.
   - Update per-model README solution notes only after a solver-backed result is reproduced.
   - Feed verified objective/solver metadata into #105 once the solution-record schema exists.

## Parallel Work Plan

Safe to work simultaneously after PR #58 and the first #55 cleanup are in place:
- Lane A: column/domain/bounds models: #68/#44 `kaibel` and #66 `gdp_col`.
- Lane B: large gas/network models: #76 `stranded_gas`, #72 `modprodnet`, and #77 `syngas`.
- Lane C: process-model semantics: #62 `biofuel`, #65 `ex1_linan_2023`, #67 `hda`, and #70/#42/#103 `methanol`.
- Lane D: smaller GDPopt strategy cleanups: #61 `batch_processing`, #63 `cstr`, #69 `med_term_purchasing`, #71 `mod_hens`, #73 `positioning`, #74 `small_batch`, and #75 `spectralog`.

Do not parallelize changes to shared benchmark infrastructure, root README generation, or common test helpers without explicit coordination. Model PRs should mostly touch one `gdplib/<model>/` package, its README, and focused tests.

## Strategy Checklist

Checked entries are strategies reported as working or completed in PR #58. Unchecked entries remain model-fix work.

- #61 `batch_processing`
  - [x] Big-M
  - [x] Hull
  - [x] Enumerate
  - [ ] GLOA
  - [x] LOA
  - [x] LBB
  - [x] RIC
- #62 `biofuel`
  - [x] Big-M
  - [ ] Hull
  - [ ] Enumerate
  - [ ] GLOA
  - [x] LOA
  - [ ] LBB
  - [x] RIC
- #63 `cstr`
  - [x] Big-M
  - [x] Hull
  - [x] Enumerate
  - [ ] GLOA
  - [x] LOA
  - [x] LBB
  - [x] RIC
- #64 `disease_model`
  - [x] Big-M
  - [x] Hull
  - [ ] Enumerate
  - [x] GLOA
  - [x] LOA
  - [x] LBB
  - [x] RIC
- #65 `ex1_linan_2023`
  - [x] Big-M
  - [ ] Hull
  - [ ] GLOA
  - [ ] LOA
  - [ ] LBB
  - [ ] RIC
- #66 `gdp_col`
  - [x] Big-M
  - [ ] Hull
  - [x] Enumerate
  - [ ] GLOA
  - [x] LOA
  - [ ] LBB
  - [x] RIC
- #67 `hda`
  - [x] Big-M
  - [ ] Hull
  - [x] Enumerate
  - [ ] GLOA
  - [x] LOA
  - [x] LBB
  - [x] RIC
- #68 `kaibel`
  - [ ] Big-M
  - [ ] Hull
  - [ ] Enumerate
  - [ ] GLOA
  - [ ] LOA
  - [ ] LBB
  - [ ] RIC
- #69 `med_term_purchasing`
  - [x] Big-M
  - [x] Hull
  - [ ] Enumerate
  - [x] GLOA
  - [x] LOA
  - [ ] LBB
  - [x] RIC
- #70 `methanol`
  - [x] Big-M
  - [x] Hull
  - [x] Enumerate
  - [ ] GLOA
  - [x] LOA
  - [x] LBB
  - [x] RIC
- #71 `mod_hens`
  - [x] Big-M
  - [x] Hull
  - [ ] Enumerate
  - [ ] GLOA
  - [x] LOA
  - [ ] LBB
  - [x] RIC
- #72 `modprodnet`
  - [x] Big-M
  - [x] Hull
  - [x] Enumerate
  - [ ] GLOA
  - [x] LOA
  - [x] LBB
  - [x] RIC
- #73 `positioning`
  - [x] Big-M
  - [x] Hull
  - [x] Enumerate
  - [ ] GLOA
  - [x] LOA
  - [x] LBB
  - [x] RIC
- #74 `small_batch`
  - [x] Big-M
  - [x] Hull
  - [x] Enumerate
  - [ ] GLOA
  - [x] LOA
  - [x] LBB
  - [x] RIC
- #75 `spectralog`
  - [x] Big-M
  - [x] Hull
  - [ ] Enumerate
  - [ ] GLOA
  - [x] LOA
  - [x] LBB
  - [x] RIC
- #76 `stranded_gas`
  - [x] Big-M
  - [ ] Hull
  - [ ] Enumerate
  - [ ] GLOA
  - [ ] LOA
  - [ ] LBB
  - [ ] RIC
- #77 `syngas`
  - [x] Big-M
  - [x] Hull
  - [x] Enumerate
  - [ ] GLOA
  - [x] LOA
  - [ ] LBB
  - [x] RIC

## Acceptance Criteria

This umbrella issue can close when:
- every per-model issue #61 through #77 is closed or explicitly split into a narrower follow-up;
- each affected model imports and builds without optional solver requirements;
- changed transformation paths have focused tests;
- solver-backed benchmark runs are documented in the relevant issue or PR with enough metadata to reproduce;
- README solution statements are updated only for verified values;
- no generated benchmark artifacts are tracked in git.

**Comment by bernalde** [comment]
# PR58 Big-M local/global benchmark evidence

| Instance | Issue | Local term | Local obj | Global term | Global obj | Warning note |
|---|---:|---|---:|---|---:|---|
| batch_processing | #61 | optimal | 679365.323409 | optimal | 679365.334869 | none |
| biofuel | #62 | intermediateNonInteger | -1.1311e+06 | optimal | 4281.529989 | none |
| cstr | #63 | optimal | 3.130197 | optimal | 3.062015 | none |
| disease_model | #64 | failure |  | optimal | 304.416211 | none |
| ex1_linan_2023 | #65 | optimal | -0.66021 | optimal | -0.9996 | none |
| gdp_col | #66 | optimal | 20081.396823 | optimal | 19740.340797 | deprecations |
| hda | #67 | optimal | 5710.789875 | infeasible |  | warnings |
| kaibel | #68 | infeasible |  | maxTimeLimit |  | warnings |
| med_term_purchasing | #69 | failure |  | optimal | 6797.539702 | deprecations |
| methanol | #70 | optimal | 645.509236 | optimal | -1574.569126 | none |
| mod_hens | #71 | optimal | 106766.531074 | optimal | 106766.531592 | none |
| modprodnet | #72 | optimal | 3592.924337 | optimal | 3592.924335 | none |
| positioning | #73 | optimal | 218.562118 | optimal | -8.064136 | none |
| small_batch | #74 | optimal | 167427.656978 | optimal | 167427.657115 | deprecations |
| spectralog | #75 | optimal | 17.26329 | optimal | 12.089262 | none |
| stranded_gas | #76 | intermediateNonInteger | -81.549342 | infeasible |  | none |
| syngas | #77 | intermediateNonInteger | 2.238163 | optimal | 4669.023494 | none |

**Comment by bernalde** [comment]
**Upstream root cause fixed for one failure class**: Pyomo/pyomo#3986 (merged 2026-08-13, resolves Pyomo/pyomo#3937, tracked here as #114) fixes GDPopt LOA hardcoding minimize sense in the OA master objective. Confirmed LOA-only — GLOA/RIC/LBB pass the sense through correctly.

Impact on this tracker:
- **LOA column re-verification needed** for native-maximize models whose LOA checkmarks predate the fix: `hda` (#67), `med_term_purchasing` (#69), `modprodnet` (#72); also `multiperiod_blending` (no open issue) and `reverse_electrodialysis` (#150) when benchmarked. Pre-fix LOA on maximize models can report wrong-direction results as "optimal".
- **Workaround reverts**: `methanol` (#70) and `stranded_gas` (#76) can drop their `minimize(-profit)` convention once benchmarks run against a Pyomo containing the fix.
- Requires a Pyomo git pin until the next release; 6.9.x/6.10.0 predate the fix.
